src/tensor/storage_traits.rs

```diff
@@ -249,6 +250,44 @@ pub trait TensorFromArray<Src, S: Shape, E: Unit>: DeviceStorage {
     fn try_tensor(&self, src: Src) -> Result<Tensor<S, E, Self>, Self::Err>;
 }
 
+/// Construct tensors from rust vectors
+pub trait TensorFromVec<E: Unit>: DeviceStorage {
+    /// Create a tensor with a dynamic shape from a rust vector
+    /// ```rust
+    /// # use dfdx::prelude::*;
+    /// # let dev: Cpu = Default::default();
+    /// let _ = dev.dynamic_tensor_from_vec(vec![1.0, 2.0, 3.0, 4.0, 5.0, 6.0], [2, 3]);
+    /// ```
+    fn dynamic_tensor_from_vec<S: Shape>(&self, src: Vec<E>, shape: S) -> Tensor<S, E, Self> {
```

Comment:
Thoughts on these names?
- `tensor_from_vec_with_shape`
- `tensor_from_shaped_vec`
- `tensor_from_vec_like` (similar to `zeros_like`/`ones_like`/etc, however I do think that is less clear in this case)

I think the `with_shape` one is the most clear about how its different from the other calls of this trait, and how you call it. Usage of `dynamic` doesn't necessarily imply you have to give a shape, but I do like that it implies it should be used with runtime shapes.

Reply:
Yeah, I've gone with tensor_from_vec_with_shape, but I'm not super happy that the names are as long as they are. It's kind of a shame that rust doesn't seem to have the features necessary to have the features necessary for us to merge the TensorFromVec and TensorFromArray traits while still allowing users to use these traits with a generic `Device<E>`.

Reply:
Yes, we could do this, but doing so would make it impossible to use dev.tensor when using a generic Device, because the extra shape parameter makes it impossible to add as a subtrait of Device. Also, that code snippet would have to look something like:
```rust
let a: Tensor<Rank1<3>, ...> = dev.tensor([1.0, 2.0, 3.0]);
let a: Tensor<Rank1<3>, ...> = dev.tensor(vec![1.0, 2, 3]);
let a: Tensor<(usize, ), ...> = dev.tensor_with_shape((vec![1.0, 2.0, 3.0], (3, ));
```
to allow optionally passing in a shape.

I may be missing something, but I was unable to merge TensorFromArray and TensorFromVec while still having the resulting trait be a subtrait of Device.